### PR TITLE
Shadow Map Start

### PR DESCRIPTION
**Pull request by lilleyse**
For #2594 

Here is basic shadow mapping for glTF models, with some debug helpers.The major areas to work on next are fixing aliasing, getting shadow commands based on the light frustum, cascades/automatic light frustum fitting, support for terrain and primitives, and otherwise addressing the TODO's I left in the code.

### CODE REVIEW

**Comment by mramato** [comment]
How do you open a pull request like this without a screen shot :smile:?

**Comment by lilleyse** [comment]
shadows2

The blue lines represent the light volume.
The actual shadow map is on the bottom right.

**Comment by bagnell** [comment]
Looks good to me so far. I don't have any comments yet. I will about how this fits into the engine when it gets to that point.

@pjcozzi Do you want to do a review?

**Comment by pjcozzi** [comment]
Yes, will check this out tonight.

File: Source/Renderer/UniformState.js [hunk 1]
```diff
@@ -885,10 +906,7 @@ define([
         this._resolutionScale = canvas.width / canvas.clientWidth;
 
         var camera = frameState.camera;
-
-        setView(this, camera.viewMatrix);
-        setInverseView(this, camera.inverseViewMatrix);
-        setCamera(this, camera);
+        this.updateCamera(camera);
```
Comment (pjcozzi) [inline Source/Renderer/UniformState.js]: I haven't looked at the rest of the code yet, but does it make sense to not call this here and require the user to call it, e.g.,
```
renderFrame
  uniformState.update()
  foreach camera
    uniformState.updateCamera()
```

File: Source/Scene/Model.js [hunk 2]
```diff
@@ -556,6 +558,17 @@ define([
         // CESIUM_RTC extension
         this._rtcCenter = undefined;    // in world coordinates
         this._rtcCenterEye = undefined; // in eye coordinates
+
+        // TODO : make a material property?
```
Comment (pjcozzi) [inline Source/Scene/Model.js]: Probably not.  This may just become part of the `Primitive` interface.

File: Source/Scene/Scene.js [hunk 3]
```diff
@@ -533,6 +535,12 @@ define([
          */
         this.fog = new Fog();
 
+        /**
+         * Render shadows in the scene.
+         * @type {ShadowMap}
+         */
+        this.shadowMap = new ShadowMap(this);
```
Comment (pjcozzi) [inline Source/Scene/Scene.js]: Haven't looked at the rest yet, but does this need to be public?

File: Source/Scene/ShadowMap.js [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+
+        updateDebugShadowViewCommand(shadowMap, frameState);
+    }
+
+    // Converts from NDC space to texture space
+    var scaleBiasMatrix = new Matrix4(0.5, 0.0, 0.0, 0.5, 0.0, 0.5, 0.0, 0.5, 0.0, 0.0, 0.5, 0.5, 0.0, 0.0, 0.0, 1.0);
+    var scratchMatrix = new Matrix4();
+
+    ShadowMap.prototype.update = function(frameState) {
+        var enabled = frameState.shadowsEnabled = this.enabled;
```
Comment (pjcozzi) [inline Source/Scene/ShadowMap.js]: The hidden logic and initialization here and then again in `executeShadowMapCommands` is hard to follow.

Instead, handle all the `enabled` logic in `executeViewportCommands`.

File: Source/Scene/Scene.js (hunk 3 above)
Comment (pjcozzi) [inline Source/Scene/Scene.js]: Probably rename to `sunShadowMap`.

File: Source/Scene/ShadowMap.js [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+        this._debugCamera = undefined;
+        this._debugShadowViewCommand = undefined;
+
+        this._depthStencilTexture = undefined;
+        this._framebuffer = undefined;
+        this._size = 1024; // Width and height of the shadow map in pixels
+
+        this.renderState = RenderState.fromCache({
+            cull : {
+                enabled : true,
```
Comment (pjcozzi) [inline Source/Scene/ShadowMap.js]: You can just add a TODO, but I'm not sure that we can always backface cull here, e.g., think of a wall geometry casting a shadow.

File: Source/Scene/ShadowMap.js [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+
+        // TODO : position camera based on sun direction
+        var camera = new Camera(scene);
+        camera.frustum = frustum;
+        var centerLongitude = -1.31968;
+        var centerLatitude = 0.698874;
+        camera.lookAt(Cartesian3.fromRadians(centerLongitude, centerLatitude), new Cartesian3(0.0, 0.0, 75.0));
+        this.camera = camera;
+
+        this.clearCommand = new ClearCommand({
```
Comment (pjcozzi) [inline Source/Scene/ShadowMap.js]: A better vertical code ordering would be to move this to just under where the render state is created.

File: Source/Scene/ShadowMap.js [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+    function updateFramebuffers(shadowMap, context) {
+        var depthStencilTexture = shadowMap._depthStencilTexture;
+        var textureChanged = !defined(depthStencilTexture) || (depthStencilTexture.width !== shadowMap._size);
+        if (!defined(shadowMap.framebuffer) || textureChanged) {
+            createFramebuffers(shadowMap, context);
+        }
+    }
+
+    ShadowMap.createShadowCastProgram = function(shaderProgram) {
+        // TODO : unused right now
```
Comment (pjcozzi) [inline Source/Scene/ShadowMap.js]: Also add a `TODO` for handling when a fragment shader does discard or writes an alpha of zero.  For example, see how we generate pick shaders:

https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Renderer/ShaderSource.js#L344-L357

It's also not clear to me yet if we want to move this shader pipeline code there.

File: Source/Scene/ShadowMap.js [hunk 8]
```diff
@@ -0,0 +1,397 @@
...
+    };
+
+    ShadowMap.createReceiveShadowsVertexShader = function(vs) {
+        vs = ShaderSource.replaceMain(vs, 'czm_shadow_main');
+        vs +=
+            'varying vec3 czm_shadowMapCoordinate; \n' +
+            'void main() \n' +
+            '{ \n' +
+            '    czm_shadow_main(); \n' +
+            '    czm_shadowMapCoordinate = (czm_shadowMapMatrix * gl_Position).xyz; \n' +
```
Comment (pjcozzi) [inline Source/Scene/ShadowMap.js]: TODO: will need to adapt this for GPU RTE, e.g., when there is low and high position attributes.

File: Source/Renderer/AutomaticUniforms.js [hunk 9]
```diff
@@ -1452,6 +1452,38 @@ define([
...
+
+        /**
+         * An automatic GLSL uniform representing the sun's shadow map depth texture.
+         *
+         * @private
+         *
+         * @alias czm_shadowMapTexture
+         * @glslUniform
+         */
+        czm_shadowMapTexture : new AutomaticUniform({
```
Comment (pjcozzi) [inline Source/Renderer/AutomaticUniforms.js]: Rename to `czm_sunShadowMapTexture` and similar renames throughout.

File: Source/Scene/ShadowMap.js [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+        var width = size;
+        var height = size;
+
+        if (!defined(shadows._debugShadowViewCommand)) {
+            var fs =
+                'varying vec2 v_textureCoordinates; \n' +
+                'void main() \n' +
+                '{ \n' +
+                '    float shadow = texture2D(czm_shadowMapTexture, v_textureCoordinates).r; \n' +
+                '    gl_FragColor = vec4(shadow, shadow, shadow, 1.0); \n' +
```
Comment (pjcozzi) [inline Source/Scene/ShadowMap.js]: Up to you, but I think `vec4(vec3(shadow), 1.0);` is better style.  Don't you remember that 565 lecture!??!?!!??!

File: Source/Scene/ShadowMap.js [hunk 11]
```diff
@@ -0,0 +1,397 @@
...
+                    }
+                }),
+                appearance : new PerInstanceColorAppearance({
+                    translucent : false,
+                    flat : true
+                }),
+                asynchronous : false
+            });
+
+            //// Create the camera frustum
```
Comment (pjcozzi) [inline Source/Scene/ShadowMap.js]: Do you still need this commented out code?  It is fine to leave it until this goes into master.

**Comment by pjcozzi** [comment]
@lilleyse consider breaking #2594 into **Short-term** and **Long-term** sections and add things like "Write tests", "reference doc", etc.  Also feel free to remove anything already there that is now irrelevant.

File: Source/Renderer/UniformState.js [hunk 12]
```diff
@@ -148,6 +148,10 @@ define([
         this._resolutionScale = 1.0;
 
         this._fogDensity = undefined;
+
+        // TODO : expose these differently
```
Comment (pjcozzi) [inline Source/Renderer/UniformState.js]: I think it would be fine to just have a `ShadowMap` reference here, and then assign to it each frame when calling `UniformState.update` (pass it as a separate parameter to start, I guess).

File: Source/Renderer/UniformState.js [hunk 13]
```diff
@@ -846,6 +850,23 @@ define([
     }
 
     /**
+     * Synchronizes the frustum's state with the camera state.  This is called
+     * by the {@link Scene} when rendering to ensure that automatic GLSL uniforms
+     * are set to the right value.
+     *
+     * @param {Object} camera The camera to synchronize with.
+     */
+    UniformState.prototype.updateCamera = function(camera) {
```
Comment (pjcozzi) [inline Source/Renderer/UniformState.js]: @bagnell so far is this inline with your vision of having separate viewports with separate cameras for things like picture-in-picture?

File: Source/Scene/Model.js (hunk 2 above)
Comment (pjcozzi) [inline Source/Scene/Model.js]: Why are these private?  Why not just `DOC_TBA`?  The user may want to override these, e.g., only have the 3D model that the camera is following cast a shadow for performance.

File: Source/Scene/Scene.js [hunk 14]
```diff
@@ -1632,6 +1640,46 @@ define([
...
+                    // TODO : Use the shader as-is for now since color mask is false. Later will need a specific shadow cast shader.
+                    //var shaderProgram = shadowMap.createShadowCastProgram(command.shaderProgram);
+                    //executeCommand(command, scene, context, passState, renderState, shaderProgram);
+                    executeCommand(command, scene, context, passState, renderState);
+                }
+            }
+        }
+
+        // Switch back to the scene's camera
+        uniformState.updateCamera(cameraScene);
```
Comment (pjcozzi) [inline Source/Scene/Scene.js]: As I mentioned in another comment, instead of this set/restore paradigm, I would rather the start of each pass just have to set this.

File: Source/Scene/Scene.js (hunk 3 above)
Comment (pjcozzi) [inline Source/Scene/Scene.js]: Passing `this` here feels wrong.  The first thing I thought was that `Scene` and `ShadowMap` were going to be badly tightly coupled.  Perhaps `new ShadowMap(context, new Camerea(this))` is better.

File: Source/Scene/ShadowMap.js [hunk 15]
```diff
@@ -0,0 +1,397 @@
...
+
+        this.passState = new PassState(scene.context);
+
+        // TODO : determine frustum based on scene
+        var frustum = new OrthographicFrustum();
+        frustum.left = -15.0;
+        frustum.right = 15.0;
+        frustum.bottom = -15.0;
+        frustum.top = 15.0;
+        frustum.near = 1.0;
```
Comment (pjcozzi) [inline Source/Scene/ShadowMap.js]: Wishful thinking speaking here: z-fighting is due to the perspective divide creating the non-linear distribution of z values, maybe we will have fewer (or just less) z-fighting issues with this orthographic projection and not need multiple frustums, other than cascades.

File: Source/Scene/ShadowMap.js [hunk 16]
```diff
@@ -0,0 +1,397 @@
...
+        Camera,
+        CullFace,
+        OrthographicFrustum,
+        Pass,
+        PerInstanceColorAppearance,
+        Primitive) {
+    "use strict";
+
+    // TODO : doc
+    // TODO : handle if depth texture extension is not supported
```
Comment (pjcozzi) [inline Source/Scene/ShadowMap.js]: If the extension is widely supported (including mobile and IE), I am fine with just requiring it.

File: Source/Scene/ShadowMap.js [hunk 17]
```diff
@@ -0,0 +1,397 @@
...
+            context : context,
+            width : shadowMap._size,
+            height : shadowMap._size,
+            pixelFormat : PixelFormat.DEPTH_STENCIL,
+            pixelDatatype : PixelDatatype.UNSIGNED_INT_24_8
+        });
+    }
+
+    function createFramebuffers(shadowMap, context) {
+        destroyTextures(shadowMap);
```
Comment (pjcozzi) [inline Source/Scene/ShadowMap.js]: This function's name is misleading.  Since it is only called once, I would move the two destroy calls and `createTextures` call to there.

**Comment by pjcozzi** [comment]
Great start, @lilleyse!  That's all my comments.

File: Source/Renderer/UniformState.js (hunk 13 above)
Comment (bagnell) [inline Source/Renderer/UniformState.js]: Yup, this is fine with me.

File: Source/Scene/Scene.js (hunk 3 above)
Comment (lilleyse) [inline Source/Scene/Scene.js]: > Haven't looked at the rest yet, but does this need to be public?

I made it public for `Shadows.html`.

Comment (pjcozzi) [inline Source/Scene/Scene.js]: Yeap, OK for now for sure.

File: Source/Scene/ShadowMap.js (hunk 11 above)
Comment (lilleyse) [inline Source/Scene/ShadowMap.js]: I will need it back, I just commented it temporarily because it wasn't working.

File: Source/Scene/ShadowMap.js (hunk 16 above)
Comment (lilleyse) [inline Source/Scene/ShadowMap.js]: Unfortunately not in IE.

File: Source/Scene/ShadowMap.js (hunk 17 above)
Comment (lilleyse) [inline Source/Scene/ShadowMap.js]: `destroyTextures` and `destroyFramebuffers` are also called in `destroy`.

I modeled the functions after `GlobeDepth`. I'm not sure what a better name would be.

Comment (lilleyse) [inline Source/Scene/ShadowMap.js]: Maybe everything can be split into `createResources` and `destroyResources` type functions.

Comment (pjcozzi) [inline Source/Scene/ShadowMap.js]: I just mean move the first three function calls in this function to the right above where `createFramebuffers` is called.

Comment (lilleyse) [inline Source/Scene/ShadowMap.js]: Ah I get it now.

File: Source/Renderer/UniformState.js (hunk 1 above)
Comment (lilleyse) [inline Source/Renderer/UniformState.js]: Removing `this.updateCamera(camera);` breaks a lot of tests, so I'm leaving it in.

**Comment by lilleyse** [comment]
Updated.

File: Source/Renderer/UniformState.js (hunk 1 above)
Comment (pjcozzi) [inline Source/Renderer/UniformState.js]: Add a TODO to your roadmap since those tests should probably call `update` and `updateCamera` (or we can make a tests-specific helper that calls both).  I hate to make the render workflow more complicated so we don't have to change tests.

**Comment by pjcozzi** [comment]
Good start!
